### PR TITLE
BN pairing precompile cost estimator

### PR DESCRIPTION
**Pull request by kilic**
#400

### CODE REVIEW

**Comment by kilic** [comment]
Also related with #351

**Review by ChihChengLiang — COMMENTED** [review]
Great work on the cost estimator and the keyless deployment!
Here's my first pass of review

File: contracts/deployment/Proxy.sol [hunk 1]
```diff
@@ -0,0 +1,69 @@
+pragma solidity ^0.5.15;
+
+import "@openzeppelin/contracts/utils/Address.sol";
+
+contract Proxy {
+    bytes32
+        private constant IMPLEMENTATION_SLOT = 0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3;
```
[the remark this answers is not in the document]
Comment (kilic) [reply contracts/deployment/Proxy.sol:7]: Sure need to document it. It's just a random 32 bytes number

File: contracts/libs/BLS.sol [hunk 2]
```diff
@@ -34,6 +37,10 @@ library BLS {
     // prettier-ignore
     uint256 private constant MASK24 = 0xffffffffffffffffffffffffffffffffffffffffffffffff;
 
+    // estimator address
+    address
+        private constant COST_ESTIMATOR_ADDRESS = 0x181f9Ced8f72AB7a8917E4D0a35084288492fC7f;
```
[the remark this answers is not in the document]
Comment (kilic) [reply contracts/libs/BLS.sol:42]: >  Do we need a specific salt to get this exact address?

Yes. btw the salt can be frontrunned we may want to hide it or authenticate it.

File: test/fast/deployment.test.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        let result = await deploy(simpleStorageFactory, salt, "SimpleStorage", verbose);
+        assert.isFalse(result.alreadyDeployed);
+
+        result = await deploy(
+            simpleStorageFactory,
+            salt,
+            "SimpleStorage",
+            verbose
+        );
+        assert.isTrue(result.alreadyDeployed);
```
[the remark this answers is not in the document]
Comment (kilic) [reply test/fast/deployment.test.ts:64]: If deployment is run earlier it only reports _already deployed_ flag

File: ts/deployment/deploy.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    signer: Signer,
+    verbose: boolean,
+    only?: { [key: string]: boolean }
+) {
+    // 1. ...
+    await deployDeployer(signer, verbose);
+    // 2. ...
+    if (!only) {
+        // deploy all
+        await deployPairingGasEstimator(signer, verbose);
```
[the remark this answers is not in the document]
Comment (kilic) [reply ts/deployment/deploy.ts:31]: No. _deploy all_ there says us deploy all other staticly addressed contracts.

File: contracts/deployment/Proxy.sol [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    // solhint-disable-next-line func-name-mixedcase
+    function __implementation__() external view returns (address) {
+        return _implementation();
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function _willFallback() internal {}
+
+    function _fallback() internal {
+        _willFallback();
```
[the remark this answers is not in the document]
Comment (kilic) [reply contracts/deployment/Proxy.sol:36]: To enable that _willFallback function can be overriden by implementation or other Proxy contract.

**Comment by ChihChengLiang** [comment]
@kilic Can you do a quick rebase for this PR?

**Comment by ChihChengLiang** [comment]
I found a disparity in the bytecodes compiled from the simpleStorage
- local `0x60806040526000805534801561001457600080fd5b5060ab806100236000396000f3fe6080604052348015600f57600080fd5b506004361060325760003560e01c8063209652551460375780635524107714604f575b600080fd5b603d606b565b60408051918252519081900360200190f35b606960048036036020811015606357600080fd5b50356071565b005b60005490565b60005556fea265627a7a723158207ccb38fe04c0aa9893716fa5bf24ce0fb0ee62b1f7b596cfb38387eb77c8c36364736f6c634300050f0032`
- CI `0x60806040526000805534801561001457600080fd5b5060ab806100236000396000f3fe6080604052348015600f57600080fd5b506004361060325760003560e01c8063209652551460375780635524107714604f575b600080fd5b603d606b565b60408051918252519081900360200190f35b606960048036036020811015606357600080fd5b50356071565b005b60005490565b60005556fea265627a7a72315820d767988a9010e191aac51c4286ce66787cc4882c8985d8a35d51c3de773f63ca64736f6c634300050f0032`

- disparity happened at local `7ccb38fe04c0aa9893716fa5bf24ce0fb0ee62b1f7b596cfb38387eb77c8c363` and CI `d767988a9010e191aac51c4286ce66787cc4882c8985d8a35d51c3de773f63ca`

I tried to put the local deployed bytecode to the online decompiler. https://ethervm.io/decompile
``` 0x6080604052348015600f57600080fd5b506004361060325760003560e01c8063209652551460375780635524107714604f575b600080fd5b603d606b565b60408051918252519081900360200190f35b606960048036036020811015606357600080fd5b50356071565b005b60005490565b60005556fea265627a7a723158207ccb38fe04c0aa9893716fa5bf24ce0fb0ee62b1f7b596cfb38387eb77c8c36364736f6c634300050f0032
```
It shows 

```
        0076    FE    *ASSERT
	0077    A2    LOG2
	0078    65    PUSH6 0x627a7a723158
	007F    20    SHA3
	0080    7C    PUSH29 0xcb38fe04c0aa9893716fa5bf24ce0fb0ee62b1f7b596cfb38387eb77c8
	009E    C3    C3
	009F    63    PUSH4 0x64736f6c
	00A4    63    PUSH4 0x4300050f
	00A9    00    *STOP
	00AA    32    ORIGIN
```
Not sure what that means

File: ts/deployment/static.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    const _proxyBytecode =
+        "0x608060405234801561001057600080fd5b50610218806100206000396000f3fe6080604052600436106100295760003560e01c80630fd5b6ae14610033578063bd5b220214610064575b610031610097565b005b34801561003f57600080fd5b506100486100b1565b604080516001600160a01b039092168252519081900360200190f35b34801561007057600080fd5b506100316004803603602081101561008757600080fd5b50356001600160a01b03166100c0565b61009f6100af565b6100af6100aa61019a565b6101bf565b565b60006100bb61019a565b905090565b60006100ca61019a565b6001600160a01b031614610125576040805162461bcd60e51b815260206004820152601a60248201527f50726f78793a20616c726561647920696e697469616c697a6564000000000000604482015290519081900360640190fd5b6001600160a01b038116610176576040805162461bcd60e51b815260206004820152601360248201527250726f78793a207a65726f206164647265737360681b604482015290519081900360640190fd5b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c355565b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c35490565b3660008037600080366000845af43d6000803e8080156101de573d6000f35b3d6000fdfea265627a7a72315820c280737031be2c6d3b410c09f08919c2cede3c62685b4b8ae0ae9fec75a0c15f64736f6c634300050f0032";
```
Comment (ChihChengLiang) [inline ts/deployment/static.ts:57]: Comment for other reviewers: We hardcode `_proxyBytecode` here. Not using the compiled one.

File: ts/deployment/static.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
...806100206000396000f3fe6080604052600436106100295760003560e01c80630fd5b6ae14610033578063bd5b220214610064575b610031610097565b005b34801561003f57600080fd5b506100486100b1565b604080516001600160a01b039092168252519081900360200190f35b34801561007057600080fd5b506100316004803603602081101561008757600080fd5b50356001600160a01b03166100c0565b61009f6100af565b6100af6100aa61019a565b6101bf565b565b60006100bb61019a565b905090565b60006100ca61019a565b6001600160a01b031614610125576040805162461bcd60e51b815260206004820152601a60248201527f50726f78793a20616c726561647920696e697469616c697a6564000000000000604482015290519081900360640190fd5b6001600160a01b038116610176576040805162461bcd60e51b815260206004820152601360248201527250726f78793a207a65726f206164647265737360681b604482015290519081900360640190fd5b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c355565b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c35490565b3660008037600080366000845af43d6000803e8080156101de573d6000f35b3d6000fdfea265627a7a72315820c280737031be2c6d3b410c09f08919c2cede3c62685b4b8ae0ae9fec75a0c15f64736f6c634300050f0032a265627a7a723158204ba46b4927debb1c1a935a0355c06a56129992e651124407c95f35579a6cb21764736f6c634300050f0032";
```
Comment (ChihChengLiang) [inline ts/deployment/static.ts:69]: Comment for other reviewers: We hardcode _bytecodeDeployer here. Not using the compiled one.

**Review by ChihChengLiang — APPROVED** [review]
Looks good to me. We use hardcoded bytecode for proxy and deployer but we could update to the compiled one after we have the Solidity 0.6 upgrade.

**Pull request merged**